### PR TITLE
Support one off slaves on the job level for mesos slaves.

### LINKED ISSUE #48: Have an option to only allow a slave to run one job and then terminate
Is it possible to have the Jenkins plugin only spin up a slave and use it for one run of a job and then kill the slave after the run is finished so that the slave only is used once? We're testing out the containerizer features and only want a container to live for one run of a job. I've seen this feature in other Jenkins plugins where the framework checks if a job has been run on a slave and if a job has run it kills the slave and spins up another one.

### PR DESCRIPTION
**Pull request by stephenliang**
Fixes #48

Based off the JClouds Jenkins plugin's implementation. The slave, on teardown, will mark temporarily offline, and thus no longer take any jobs. The Mesos retention strategy will then set it as pending deletion after the idle period to protect against post-build actions from terminating the slave too soon. The final step is the asynchronous period work in MesosCleanupTask which will terminate the pending deletion slave.

### CODE REVIEW

File: src/main/java/org/jenkinsci/plugins/mesos/MesosCleanupThread.java [hunk 1]
```diff
@@ -0,0 +1,110 @@
...
+import com.google.common.util.concurrent.ListeningExecutorService;
+import com.google.common.util.concurrent.MoreExecutors;
+import hudson.Extension;
+import hudson.model.AsyncPeriodicWork;
+import hudson.model.Computer;
+import hudson.model.TaskListener;
+import jenkins.model.Jenkins;
+
+/**
+ This file is part of the JCloud Jenkins Plugin. (https://github.com/jenkinsci/jclouds-plugin)
```
Comment (barkerd427) [inline src/main/java/org/jenkinsci/plugins/mesos/MesosCleanupThread.java]: Could we reference the commit from where this version initiated?

File: src/main/java/org/jenkinsci/plugins/mesos/MesosCleanupThread.java [hunk 2]
```diff
@@ -0,0 +1,110 @@
...
+        final ImmutableList.Builder<ListenableFuture<?>> deletedNodesBuilder = ImmutableList.<ListenableFuture<?>>builder();
+        ListeningExecutorService executor = MoreExecutors.listeningDecorator(Computer.threadPoolForRemoting);
+        final ImmutableList.Builder<MesosComputer> computersToDeleteBuilder = ImmutableList.<MesosComputer>builder();
+
+        for (final Computer c : Jenkins.getInstance().getComputers()) {
+            if (MesosComputer.class.isInstance(c)) {
+                if (((MesosComputer) c).getNode().isPendingDelete()) {
+                    final MesosComputer comp = (MesosComputer) c;
+                    computersToDeleteBuilder.add(comp);
+                    logger.log(Level.INFO, "Marked " + comp.getName() + " for deletion");
```
Comment (barkerd427) [inline src/main/java/org/jenkinsci/plugins/mesos/MesosCleanupThread.java]: nit: String.format("Marked %s for deletion.", comp.getName())

The same for other string formatting sections.

File: src/main/java/org/jenkinsci/plugins/mesos/MesosCleanupThread.java [hunk 3]
```diff
@@ -0,0 +1,110 @@
...
+                    ListenableFuture<?> f = executor.submit(new Runnable() {
+                        public void run() {
+                            logger.log(Level.INFO, "Deleting pending node " + comp.getName());
+                            try {
+                                comp.getNode().terminate();
+                            } catch (RuntimeException e) {
+                                logger.log(Level.WARNING, "Failed to disconnect and delete " + c.getName() + ": " + e.getMessage());
```
Comment (barkerd427) [inline src/main/java/org/jenkinsci/plugins/mesos/MesosCleanupThread.java]: Can we not use `comp` at this point instead of `c`?

File: src/main/java/org/jenkinsci/plugins/mesos/MesosCleanupThread.java [hunk 4]
```diff
@@ -0,0 +1,110 @@
...
+                                throw e;
+                            }
+                        }
+                    });
+                    deletedNodesBuilder.add(f);
+                } else {
+                    logger.log(Level.INFO, c.getName() + " is not pending deletion");
```
Comment (barkerd427) [inline src/main/java/org/jenkinsci/plugins/mesos/MesosCleanupThread.java]: Should this be a DEBUG as it is the normal state? This plugin already produces a lot of INFO level output.

File: src/main/java/org/jenkinsci/plugins/mesos/MesosCleanupThread.java [hunk 5]
```diff
@@ -0,0 +1,110 @@
...
+                }
+            } else {
+                logger.log(Level.INFO, c.getName() + " is not a mesos computer, it is a " + c.getClass().getName());
```
Comment (barkerd427) [inline src/main/java/org/jenkinsci/plugins/mesos/MesosCleanupThread.java]: This could also be DEBUG.

**Comment by barkerd427** [comment]
This looks like a reasonable solution, and I'm guessing it's working for JClouds. I haven't tested functionality.

File: src/main/java/org/jenkinsci/plugins/mesos/MesosCleanupThread.java (hunk 1 above)
Comment (stephenliang) [inline src/main/java/org/jenkinsci/plugins/mesos/MesosCleanupThread.java]: Sure

File: src/main/java/org/jenkinsci/plugins/mesos/MesosCleanupThread.java (hunk 4 above)
Comment (stephenliang) [inline src/main/java/org/jenkinsci/plugins/mesos/MesosCleanupThread.java]: Changed it to FINE

File: src/main/java/org/jenkinsci/plugins/mesos/MesosCleanupThread.java (hunk 5 above)
Comment (stephenliang) [inline src/main/java/org/jenkinsci/plugins/mesos/MesosCleanupThread.java]: Changed it to FINE

File: src/main/java/org/jenkinsci/plugins/mesos/JenkinsScheduler.java [hunk 6]
```diff
@@ -407,6 +408,28 @@ private void createMesosTask(Offer offer, Request request) {
     LOGGER.info("Launching task " + taskId.getValue() + " with URI " +
                 joinPaths(jenkinsMaster, SLAVE_JAR_URI_SUFFIX));
 
+    for (final Computer computer : Jenkins.getInstance().getComputers()) {
```
Comment (vinodkone) [inline src/main/java/org/jenkinsci/plugins/mesos/JenkinsScheduler.java]: I'm not sure I follow the logic here. Why are we creating a mesos task with the same name as an existing jenkins slave in the first place?

Comment (stephenliang) [inline src/main/java/org/jenkinsci/plugins/mesos/JenkinsScheduler.java]: We are, you can see that the slave name is the name of the mesos task. It just so happens that you can be offered a task after it has been started or deleted which leads to orphan tasks and containers

On June 21, 2015 3:10:49 PM CDT, Vinod Kone <notifications@github.com> wrote:
>> @@ -407,6 +408,28 @@ private void createMesosTask(Offer offer,
>Request request) {
>>      LOGGER.info("Launching task " + taskId.getValue() + " with URI "
>+
>>                  joinPaths(jenkinsMaster, SLAVE_JAR_URI_SUFFIX));
>>  
>> +    for (final Computer computer :
>Jenkins.getInstance().getComputers()) {
>
>I'm not sure I follow the logic here. Why are we creating a mesos task
>with the same name as an existing jenkins slave in the first place? 
>
>---
>Reply to this email directly or view it on GitHub:
>https://github.com/jenkinsci/mesos-plugin/pull/121/files#r32896961

-- 
Sent from my Android device with K-9 Mail. Please excuse my brevity.

File: src/main/java/org/jenkinsci/plugins/mesos/MesosRetentionStrategy.java [hunk 7]
```diff
@@ -86,7 +86,11 @@ private long checkInternal(MesosComputer c) {
 
       if (idleMilliseconds > MINUTES.toMillis(idleTerminationMinutes)) {
         LOGGER.info("Disconnecting idle computer " + c.getName());
-        c.getNode().terminate();
+        c.getNode().setPendingDelete(true);
```
Comment (vinodkone) [inline src/main/java/org/jenkinsci/plugins/mesos/MesosRetentionStrategy.java]: Why the change here to mark it for deletion instead of directly terminate it?

Comment (stephenliang) [inline src/main/java/org/jenkinsci/plugins/mesos/MesosRetentionStrategy.java]: Because then you end up terminating a slave that's still doing things in the post build state, leading to a failed build. Rather we should wait until idle time has elapsed

On June 21, 2015 3:21:28 PM CDT, Vinod Kone <notifications@github.com> wrote:
>> @@ -86,7 +86,11 @@ private long checkInternal(MesosComputer c) {
>>  
>>        if (idleMilliseconds >
>MINUTES.toMillis(idleTerminationMinutes)) {
>>          LOGGER.info("Disconnecting idle computer " + c.getName());
>> -        c.getNode().terminate();
>> +        c.getNode().setPendingDelete(true);
>
>Why the change here to mark it for deletion instead of directly
>terminate it?
>
>---
>Reply to this email directly or view it on GitHub:
>https://github.com/jenkinsci/mesos-plugin/pull/121/files#r32897165

-- 
Sent from my Android device with K-9 Mail. Please excuse my brevity.

File: src/main/java/org/jenkinsci/plugins/mesos/MesosOneOffSlave.java [hunk 8]
```diff
@@ -0,0 +1,95 @@
...
+
+    @Override
+    @SuppressWarnings("rawtypes")
+    public Environment setUp(AbstractBuild build, Launcher launcher, final BuildListener listener) {
+        if (MesosComputer.class.isInstance(build.getExecutor().getOwner())) {
+            final MesosComputer c = (MesosComputer) build.getExecutor().getOwner();
+            return new Environment() {
+                @Override
+                public boolean tearDown(AbstractBuild build, final BuildListener listener) throws IOException, InterruptedException {
+                    LOGGER.warning("Single-use slave " + c.getName() + " getting torn down.");
```
Comment (vinodkone) [inline src/main/java/org/jenkinsci/plugins/mesos/MesosOneOffSlave.java]: Lets stick with one terminology. either "one off slave" or "single use slave" for file names, variable names and log messages.

File: src/main/java/org/jenkinsci/plugins/mesos/MesosRetentionStrategy.java (hunk 7 above)
Comment (vinodkone) [inline src/main/java/org/jenkinsci/plugins/mesos/MesosRetentionStrategy.java]: Oh. Didn't realize that a slave is considered idle if it's doing post build things!?

File: src/main/java/org/jenkinsci/plugins/mesos/JenkinsScheduler.java (hunk 6 above)
Comment (vinodkone) [inline src/main/java/org/jenkinsci/plugins/mesos/JenkinsScheduler.java]: Sorry for not being clear. Yes, the id of the mesos task is the same as the name of the Jenkins slave and this was done to make it easy to figure out which mesos task to kill when the jenkins slave goes idle.

What I am confused about is that, this code suggests we are creating duplicate task ids which is unexpected. Can you elaborate what do you mean by your second sentence?

Thanks.

Comment (stephenliang) [inline src/main/java/org/jenkinsci/plugins/mesos/JenkinsScheduler.java]: It's possible to receive an offer to launch the same mesos task after it has already been launched. There's nothing preventing that from happening right now and it leads to orphans. We've seen tasks that have launched in mesos but never get killed and don't do anything because they launch after it has already been launched. These if statements guard from that race condition

On June 21, 2015 3:38:04 PM CDT, Vinod Kone <notifications@github.com> wrote:
>> @@ -407,6 +408,28 @@ private void createMesosTask(Offer offer,
>Request request) {
>>      LOGGER.info("Launching task " + taskId.getValue() + " with URI "
>+
>>                  joinPaths(jenkinsMaster, SLAVE_JAR_URI_SUFFIX));
>>  
>> +    for (final Computer computer :
>Jenkins.getInstance().getComputers()) {
>
>Sorry for not being clear. Yes, the id of the mesos task is the same as
>the name of the Jenkins slave and this was done to make it easy to
>figure out which mesos task to kill when the jenkins slave goes idle.
>
>What I am confused about is that, this code suggests we are creating
>duplicate task ids which is unexpected. Can you elaborate what do you
>mean by your second sentence?
>
>Thanks.
>
>
>---
>Reply to this email directly or view it on GitHub:
>https://github.com/jenkinsci/mesos-plugin/pull/121/files#r32897375

-- 
Sent from my Android device with K-9 Mail. Please excuse my brevity.

File: src/main/java/org/jenkinsci/plugins/mesos/MesosOneOffSlave.java (hunk 8 above)
Comment (stephenliang) [inline src/main/java/org/jenkinsci/plugins/mesos/MesosOneOffSlave.java]: Done

File: src/main/java/org/jenkinsci/plugins/mesos/JenkinsScheduler.java (hunk 6 above)
Comment (maselvaraj) [inline src/main/java/org/jenkinsci/plugins/mesos/JenkinsScheduler.java]: Sorry, I am also not clear on the rational behind launching the task again. I was wondering if we need to skip this task launch, if it is pending delete ?

Comment (stephenliang) [inline src/main/java/org/jenkinsci/plugins/mesos/JenkinsScheduler.java]: Actually we should remove pending delete if it's done launching. That's in the code now, a change was made to prevent a slave from starting and then deleted mid-test

**Comment by maselvaraj** [comment]
@stephenliang @vinodkone I hope all the review comments have been addressed ? 
I just have a concern on this line https://github.com/stephenliang/mesos-plugin/commit/4475bca434c0b3843534c8d5bd775c4348000d35#diff-2a5f1ead1e1896276364ab8fcf73578bR444
Can you help me understand what does it do? Trying to launch a dummy task so that it doesn't get launched again and again ?

**Comment by stephenliang** [comment]
I was basically declining it. It's duplicate of declineOffer so I changed it to do that

**Comment by maselvaraj** [comment]
Got it! Have you addressed all other comments ? If yes, can you please merge the latest changes from master?

**Comment by stephenliang** [comment]
Okay, done

**Comment by maselvaraj** [comment]
Thank you @stephenliang !

File: pom.xml [hunk 9]
```diff
@@ -125,6 +126,11 @@
           <version>${assertj.version}</version>
           <scope>test</scope>
       </dependency>
+      <dependency>
+          <groupId>com.google.guava</groupId>
+          <artifactId>guava</artifactId>
+          <version>${guava.version}</version>
```
Comment (jglick) [inline pom.xml]: If you are not using plugin-first class loading or classpath masking, this is wrong. [Details](https://jenkins.io/doc/developer/plugin-development/dependencies-and-class-loading/)

Comment (jtnord) [reply pom.xml]: #322
